src/java/org/apache/cassandra/index/sai/utils/IndexFileUtils.java

```diff
@@ -109,11 +112,49 @@ public interface ChecksumWriter
 
     class IncrementalChecksumSequentialWriter extends SequentialWriter implements ChecksumWriter
     {
-        private final CRC32 checksum = new CRC32();
-
-        IncrementalChecksumSequentialWriter(File file)
+        /** Protects the checksum so only one Writer can update it */
+        private Guard<FileChecksum> checksumGuard;
+        /** Current (running) checksum from the beginning of the file till the current position */
+        private FileChecksum checksum;
+        /** Remembers the checksum after closing this writer */
+        private long finalChecksum;
+
+        IncrementalChecksumSequentialWriter(File file, boolean append) throws IOException
         {
             super(file, writerOption);
+
+            while (checksum == null)
+            {
+                checksumGuard = checksumCache.get(file.path(), s -> new Guard<>(new FileChecksum()));
+                checksum = checksumGuard.tryLock();
+
+                if (checksum == null)
+                    // If we're here this means some other Writer did not unlock the checksum object,
+                    // so we can't use the same checksum safely, as there is a slight probability it
+                    // is in active use. This is not necessarily a bug - e.g. it is also possible
+                    // the other writer was interrupted and the client code simply forgot to close() it.
+                    // Therefore, we'll get a new one, just to be safe.
+                    checksumCache.invalidate(file.path());
```

Comment:
should we have some warning logs here, so we can debug in the future if it ever happens.

Reply:
+1